### PR TITLE
FOUNDATIONS: Add VideoMetadata Test

### PR DESCRIPTION
**Pull request by UmidjonDeveloperr**

### CODE REVIEW

**Comment by ZafarUrakov** [comment]
approved

File: WatchWave.Api.Tests.Unit/Services/Foundations/VideoMetadatas/VideoMetadataServiceTests.Logic.Add.cs [hunk 1]
```diff
@@ -0,0 +1,43 @@
+﻿using FluentAssertions;
```
Comment (ElbekDeveloper) [inline WatchWave.Api.Tests.Unit/Services/Foundations/VideoMetadatas/VideoMetadataServiceTests.Logic.Add.cs]: Header

File: WatchWave.Api.Tests.Unit/Services/Foundations/VideoMetadatas/VideoMetadataServiceTests.Validation.Add.cs [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+namespace WatchWave.Api.Tests.Unit.Services.Foundations.VideoMetadatas
+{
+	public partial class VideoMetadataServiceTests
+	{
+		[Fact]
+		public async Task ShouldThrowValidationexceptionOnAddIfVideoMetadataIsNullAndLogErrorAsync()
+		{
+			//given
+			VideoMetadata nullVideoMetadata = null;
```
Comment (ElbekDeveloper) [inline WatchWave.Api.Tests.Unit/Services/Foundations/VideoMetadatas/VideoMetadataServiceTests.Validation.Add.cs:19]: <Nullable>disalbe</Nullable>

File: WatchWave.Api/Brokers/Storages/IStorageBroker.VideoMetadatas.cs [hunk 3]
```diff
@@ -7,7 +7,7 @@
 
 namespace WatchWave.Api.Brokers.Storages
 {
-    internal partial interface IStorageBroker
+    public partial interface IStorageBroker
```
Comment (ElbekDeveloper) [inline WatchWave.Api/Brokers/Storages/IStorageBroker.VideoMetadatas.cs]: internal partial

File: WatchWave.Api/Models/VideoMetadatas/Exceptions/VideoMetadataValidationException.cs [hunk 4]
```diff
@@ -0,0 +1,12 @@
+﻿using Xeptions;
+
+namespace WatchWave.Api.Models.VideoMetadatas.Exceptions
+{
+	public class VideoMetadataValidationException : Xeption
+	{
+        public VideoMetadataValidationException(Exception innerException)
+            :base(message: "VideoMetadata Validation Exception occured, fix the errors and try again.",
+                 innerException)
+        { }
```
Comment (ElbekDeveloper) [inline WatchWave.Api/Models/VideoMetadatas/Exceptions/VideoMetadataValidationException.cs]: Bu eski usul. Yangisida qiymat bo'lmaydi

**Review by ElbekDeveloper — CHANGES_REQUESTED** [review]

File: WatchWave.Api/Brokers/Storages/StorageBroker.VideoMetadatas.cs [hunk 5]
```diff
@@ -8,7 +8,7 @@
 
 namespace WatchWave.Api.Brokers.Storages
 {
-    internal partial class StorageBroker
+    public partial class StorageBroker
```
Comment (ElbekDeveloper) [inline WatchWave.Api/Brokers/Storages/StorageBroker.VideoMetadatas.cs:11]: Internal

File: WatchWave.Api/Brokers/Storages/StorageBroker.cs [hunk 6]
```diff
@@ -8,7 +8,7 @@
 
 namespace WatchWave.Api.Brokers.Storages
 {
-    internal partial class StorageBroker : EFxceptionsContext, IStorageBroker
+    public partial class StorageBroker : EFxceptionsContext, IStorageBroker
```
Comment (ElbekDeveloper) [inline WatchWave.Api/Brokers/Storages/StorageBroker.cs:11]: Internal

Comment (ElbekDeveloper) [reply WatchWave.Api/Brokers/Storages/StorageBroker.cs:11]: [Buni ko'ring](https://www.youtube.com/watch?v=dweVExK8IPs)

File: WatchWave.Api/Models/VideoMetadatas/Exceptions/AlreadyExistVideoMetadataException.cs [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+// Free To Use To Find Comfort and Peace
+//==================================================
+
+using Xeptions;
+
+namespace WatchWave.Api.Models.VideoMetadatas.Exceptions
+{
+	public class AlreadyExistVideoMetadataException : Xeption
+	{
+        public AlreadyExistVideoMetadataException(string message, Exception innerException)
```
Comment (ElbekDeveloper) [inline WatchWave.Api/Models/VideoMetadatas/Exceptions/AlreadyExistVideoMetadataException.cs]: AlreadyExistsVideoMetadataException

File: WatchWave.Api/Services/VideoMetadatas/VideoMetadataService.Exceptions.cs [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+
+using Microsoft.Data.SqlClient;
+using STX.EFxceptions.Abstractions.Models.Exceptions;
+using WatchWave.Api.Models.VideoMetadatas;
+using WatchWave.Api.Models.VideoMetadatas.Exceptions;
+using Xeptions;
+
+namespace WatchWave.Api.Services.VideoMetadatas
+{
+	public partial class VideoMetadataService
```
Comment (ElbekDeveloper) [inline WatchWave.Api/Services/VideoMetadatas/VideoMetadataService.Exceptions.cs:15]: Internal

File: WatchWave.Api/Services/VideoMetadatas/VideoMetadataService.Validations.cs [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+			Condition = string.IsNullOrWhiteSpace(text),
+			Message = "Text is required."
+		};
+
+		private static dynamic IsInvalid(DateTimeOffset date) => new
+		{
+			Condition = date == default(DateTimeOffset),
+			Message = "Date is required."
+		};
+
```
Comment (ElbekDeveloper) [inline WatchWave.Api/Services/VideoMetadatas/VideoMetadataService.Validations.cs:81]: CreatedDate != Updated

Comment (ElbekDeveloper) [reply WatchWave.Api/Services/VideoMetadatas/VideoMetadataService.Validations.cs:81]: CreatedDate is not recent

File: WatchWave.Api/Services/VideoMetadatas/VideoMetadataService.cs [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+		public ValueTask<VideoMetadata> AddVideoMetadataAsync(VideoMetadata videoMetadata) =>
+			TryCatch(async () =>
+			{
+				ValidateVideoMetadataOnAdd(videoMetadata);
+
+				return await this.storageBroker.InsertVideoMetadataAsync(videoMetadata);
+			});
+
+
+
```
Comment (ElbekDeveloper) [inline WatchWave.Api/Services/VideoMetadatas/VideoMetadataService.cs]: Bu ortiqcha qatorlar

File: WatchWave.Api/Services/VideoMetadatas/VideoMetadataService.cs [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+			this.storageBroker = storageBroker;
+			this.loggingBroker = loggingBroker;
+		}
+
+		public ValueTask<VideoMetadata> AddVideoMetadataAsync(VideoMetadata videoMetadata) =>
+			TryCatch(async () =>
+			{
+				ValidateVideoMetadataOnAdd(videoMetadata);
+
+				return await this.storageBroker.InsertVideoMetadataAsync(videoMetadata);
```
Comment (ElbekDeveloper) [inline WatchWave.Api/Services/VideoMetadatas/VideoMetadataService.cs:34]: Bir tab orqaga

File: WatchWave.Api.Tests.Unit/Services/Foundations/VideoMetadatas/VideoMetadataServiceTests.Exception.Add.cs [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+			string someString = GetRandomString();
+
+			var duplicateKeyException = new DuplicateKeyException(someString);
+
+			AlreadyExistVideoMetadataException alreadyExistVideoMetadataException =
+				new("Video Metadata already exist, please try again.",
+					duplicateKeyException);
+
+			VideoMetadataDependencyValidationException expectedVideoMetadataDependencyValidationException
+				= new("Video Metadata dependency error occured. Fix errors and try again.",
```
Comment (ElbekDeveloper) [inline WatchWave.Api.Tests.Unit/Services/Foundations/VideoMetadatas/VideoMetadataServiceTests.Exception.Add.cs]: Bu standard emas

File: WatchWave.Api.Tests.Unit/Services/Foundations/VideoMetadatas/VideoMetadataServiceTests.Exception.Add.cs [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+			VideoMetadataDependencyValidationException actualVideoMetadataDependencyValidationException =
+				await Assert.ThrowsAnyAsync<VideoMetadataDependencyValidationException>(addVideoMetadataTask.AsTask);
+
+			//then
+			actualVideoMetadataDependencyValidationException.Should().BeEquivalentTo(
+				expectedVideoMetadataDependencyValidationException);
+
+			this.storageBrokerMock.Verify(broker =>
+				broker.InsertVideoMetadataAsync(It.IsAny<VideoMetadata>()),
+					Times.Once());
```
Comment (ElbekDeveloper) [inline WatchWave.Api.Tests.Unit/Services/Foundations/VideoMetadatas/VideoMetadataServiceTests.Exception.Add.cs:96]: Qavs kerakmas

File: WatchWave.Api.Tests.Unit/Services/Foundations/VideoMetadatas/VideoMetadataServiceTests.Validation.Add.cs [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+				values: "Id is required.");
+
+			invalidVideoMetadataException.AddData(key: nameof(VideoMetadata.Title),
+				values: "Text is required.");
+
+			invalidVideoMetadataException.AddData(key: nameof(VideoMetadata.BlobPath),
+				values: "Text is required.");
+
+			invalidVideoMetadataException.AddData(key: nameof(VideoMetadata.CreatedDate),
+				values: "Date is required.");
```
Comment (ElbekDeveloper) [inline WatchWave.Api.Tests.Unit/Services/Foundations/VideoMetadatas/VideoMetadataServiceTests.Validation.Add.cs]: Buni formatini to'g'rilang

**Comment by ElbekDeveloper** [comment]
`DbUpdateException` ham handle qilinsin

**Pull request merged**